lib/screens/v2/widget_instance/dup_alert.ex

```diff
@@ -0,0 +1,299 @@
+  def slot_names(t) do
+    base_slot_name =
+      case alert_layout(t) do
+        :full_screen -> "full_rotation"
+        :partial -> "bottom_pane"
+        # This case will never match since it would be filtered out at `valid_candidate?`,
+        # but Dialyzer doesn't know that.
+        :no_render -> "bottom_pane"
+      end
+
+    # Returns e.g. [:full_rotation_zero], [:bottom_pane_one], ...
+    [:"#{base_slot_name}_#{t.rotation_index}"]
+  end
+
+  @spec widget_type(t()) :: :takeover_alert | :partial_alert
+  def widget_type(%__MODULE__{} = t) do
+    case alert_layout(t) do
+      :full_screen -> :takeover_alert
+      :partial -> :partial_alert
+      # This case will never match since it would be filtered out at `valid_candidate?`,
+      # but Dialyzer doesn't know that.
+      :no_render -> :partial_alert
+    end
+  end
+
+  def valid_candidate?(%__MODULE__{} = t) do
+    alert_layout(t) != :no_render
+  end
+
+  @spec alert_layout(t()) :: :full_screen | :partial | :no_render
+  defp alert_layout(t) do
+    effect = t.alert.effect
+    location = BaseAlert.location(t)
+    affected_line_count = length(get_affected_lines(t))
+    total_line_count = length(get_total_lines(t))
+    rotation_index = t.rotation_index
+
+    parameters = {effect, location, affected_line_count, total_line_count, rotation_index}
+
+    delay_severity = t.alert.severity
+
+    if effect == :delay and delay_severity < 5 do
+      # We don't show delays < 20 minutes
+      log_layout_mismatch(t, parameters, delay_severity)
+      :no_render
+    else
+      case do_alert_layout(parameters) do
+        :no_render ->
+          log_layout_mismatch(t, parameters, delay_severity)
+          :no_render
+
+        layout ->
+          layout
+      end
+    end
+  end
+
+  # The widget _should_ always be valid given the constraints of what alerts
+  # we generate DUP alert widgets for, but in case one slips through, we
+  # should know about it!
+  defp log_layout_mismatch(t, parameters, delay_severity) do
+    {effect, location, affected_line_count, total_line_count, rotation_index} = parameters
+
+    labeled_parameters = [
+      effect: effect,
+      delay_severity: delay_severity,
+      location: location,
+      affected_line_count: affected_line_count,
+      total_line_count: total_line_count,
+      rotation_index: rotation_index
+    ]
+
+    identifiers = [alert_id: t.alert.id, stop_ids: Enum.join(t.screen.alerts.stop_ids, ",")]
+    log_fields = identifiers ++ labeled_parameters
+
+    Logger.info(
+      "[DUP alert no matching layout] " <>
+        Enum.map_join(log_fields, " ", fn {label, value} -> "#{label}=#{value}" end)
+    )
+  end
+
+  # Inputs/outputs are stored as a table for readability.
+  # Table is adapted from the one in the product spec: https://www.notion.so/mbta-downtown-crossing/DUP-Alert-Widget-Specification-a82acff850ed4f2eb98a04e5f3e0fe52
+  # Compile-time code converts each table row to `do_alert_layout` function clauses.
```

Comment:
I agree it turned out a bit more brittle than I wanted when I started out, since it's not easy to generate the appropriate clauses for values of "any" or "any boundary" in the table, forcing me to use explicit location values and duplicate rows.

However I still prefer keeping the conditions in a table like this for conciseness and readability. While working on this, I was able to be confident that the code follows the spec because the logic is encoded here almost exactly the same as it is in the spec.

> Let's say we had new cases where a particular station is an exception, so we want a column with station name for only that one case. We could easily do that with a new pattern matching function, but then we'll have some cases in the table, and some cases written separately.

This is true, but the `delay` cases are already handled separately with a hand-written clause. Not necessarily a good thing, just saying this already happens. Maybe it would be enough to add a "comment" line to the table for each special case that's handled separately?

Reply:
Sorry, I meant the `delay` case was already mentioned in a comment ([here](https://github.com/mbta/screens/pull/1634/files#diff-97879ae25268a9f3b56c913f936c95193ad09831de6b1b5b764c35e74372f7eaR162)), I was just talking about what we'd do for your hypothetical case where we'd need to handle a specific exceptional station.